runtime/ini.cpp

```diff
@@ -0,0 +1,240 @@
+const int64_t ALL_QUOTES = 2;
+
+string clear_quotes(const string &str, int64_t flag) {
+  if (str.empty()) {
+    return {};
+  }
+
+  Optional<string> clear_str;
+
+  switch (flag) {
+    case SINGLE_QOUTES:
+        clear_str = f$preg_replace(string("/'+/"), string(""), str);
+      break;
+    case DOUBLE_QUOTES:
+        clear_str = f$preg_replace(string("/\"+/"), string(""), str);
+      break;
+    case ALL_QUOTES:
+        clear_str = f$preg_replace(string("/['|\"]+/"), string(""), str);
+      break;
+  }
+
+  if (clear_str.is_null()) {
+    return string("");
+  }
+
+  return clear_str.ref();
+}
+
+string clear_extra_spaces(const string &str) {
+  if (str.empty()) {
+    return {};
+  }
+
+  Optional<string> clear_str;
+
+  clear_str = f$preg_replace(string("/ +/"), string(" "), str);
+
+  if (clear_str.is_null()) {
+    return string("");
+  }
+
+  return clear_str.ref();
+}
+
+/*
+ * INI parsing functions
+ */
+
+bool is_ini_section(const string &ini_section) {
+  return f$preg_match(string("/^\\[.+\\]$/"), ini_section).val();
+}
+
+bool is_ini_var(const string &ini_var) {
+  return f$preg_match(string("/^.+/"), ini_var).val();
+}
+
+bool is_ini_val(const string &ini_val) {
+  return f$preg_match(string("/(.*\n(?=[A-Z])|.*$)/"), ini_val).val();
+}
+
+bool is_ini_bool_val(const string &ini_val) {
+  return f$preg_match(string("/^(false|true|0|1|off|on)$/i"), ini_val).val();
+}
+
+bool is_ini_float_val(const string &ini_val) {
+  return f$preg_match(string(R"(/^\d+\.\d*$/)"), ini_val).val();
+}
+
+bool is_ini_str_val(const string &ini_val) {
+  return f$preg_match(string("/^.*$/i"), ini_val).val();
+}
+
+string get_ini_section(const string &ini_entry) {
+  if (ini_entry.empty()) {
+    return {};
+  }
+
+  return ini_entry.substr(1, ini_entry.size()-2);
+}
+
+array<string> split_ini_entry(const string &ini_entry) {
+  if (ini_entry.empty()) {
+    return {};
+  }
+
+  array<string> res = f$explode(string("="), ini_entry, 2);
+
+  if (res.size().int_size != 2) {
+    php_warning("Error");
+    return {};
+  }
+
+  return res;
+}
+
+bool cast_str_to_bool(const string &ini_var) {
+  if (ini_var.empty()) {
+    return false;
+  }
+
+  string ini_bool_var = f$strtolower(ini_var);
+
+  if (ini_bool_var == string("on") || ini_bool_var == string("1") || ini_bool_var == string("true")) {
+    return true;
+  }
+
+  if (ini_bool_var == string("off") || ini_bool_var == string("0") || ini_bool_var == string("false")) {
+    return false;
+  }
+
+  return false;
+}
+
+array<mixed> f$parse_ini_string(const string &ini_string, bool process_sections, int scanner_mode) {
+  if (ini_string.empty()) {
+    return {};
+  }
+
+  string ini_string_copy = f$trim(ini_string);
+  ini_string_copy = clear_extra_spaces(ini_string_copy);
+
+  array<mixed> res(array_size(0, 0, true));
+  array<mixed> section(array_size(0, 0, true));
+  string ini_entry;
+  string ini_section;
+
+  for (string::size_type i = 0; i <= ini_string_copy.size(); ++i) {
```

Comment:
Firstly, also it's better to use `auto` here.
Secondly, it looks very strange that your `for` goes to `ini_string_copy.size()`, suddenly there will be a situation that you turn to `ini_string_copy[ini_string_copy.size()])` (for example, if a string of the form `[text` somehow gets to your function). It's not very good.
It's better to fix on 
```c++
for (auto i = 0; i < ini_string_copy.size(); ++i) {
...
}
```

Reply:
Type of `i` is deduced as `int`, it is not an intention.

Reply:
> Type of `i` is deduced as `int`, it is not an intention.

Thanks, you're right.
Fixed